cpp/src/jni/orc/concurrent_map.h

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ */
+
+#ifndef JNI_ID_TO_MODULE_MAP_H
+#define JNI_ID_TO_MODULE_MAP_H
+
+#include <memory>
+#include <mutex>
+#include <unordered_map>
+#include <utility>
+
+#include "arrow/util/macros.h"
+
+namespace arrow {
+namespace jni {
+
+/**
+ * An utility class that map module id to module pointers.
+ * @tparam Holder class of the object to hold.
+ */
+template <typename Holder>
+class concurrentMap {
+ public:
+  concurrentMap() : module_id_(kInitModuleId) {}
+
+  jlong Insert(Holder holder) {
+    std::lock_guard<std::mutex> lock(mtx_);
+    jlong result = module_id_++;
+    map_.insert(std::pair<jlong, Holder>(result, holder));
+    return result;
+  }
+
+  void Erase(jlong module_id) {
+    std::lock_guard<std::mutex> lock(mtx_);
+    map_.erase(module_id);
+  }
+
+  Holder Lookup(jlong module_id) {
+    Holder result = NULLPTR;
+    try {
+      result = map_.at(module_id);
+    } catch (const std::out_of_range& e) {
+    }
+    if (result != NULLPTR) {
+      return result;
+    }
+    std::lock_guard<std::mutex> lock(mtx_);
```

Comment:
duplicate code?  Why the code duplication?

Reply:
According to this patch https://github.com/dremio/gandiva/pull/82
It would avoid lock overhead whenever possible.

Reply:
This isn't thread safe though?

Reply:
@emkornfield : @yuruiz is right we wanted to avoid locking in the execute paths as much as possible. this was a short cut to do a quick/dirty lookup to see if we can grab the module.

Reply:
@praveenbingo isn't this undefined behavior if there is a concurrent write?

Reply:
I remember there was some reasoning we used on why this was okay to do. but, I'm not able to recollect now.

@praveenbingo I think we should either fix the race or add a comment on why it's safe to do.

Reply:
As far as i remember, map.at() was a const method and c++ specification seemed to mandate that it was safe to use such methods if modifications are happening to different elements, which would be the case in this map.

It is somewhat grey to me if the specification also covers the map being resized (i.e. container being modified).

[1] https://stackoverflow.com/questions/16130494/are-stdmap-and-stdvector-thread-safe

Reply:
Modifications to different elements I believe would be safe.  I don't think this holds for insertions.  I would either like to see: 
1.  This code removed and only accessed under the lock.
2.  A unit test run under TSAN that shows correct behavior with multiple readers and writers.

If locking on the read path is a problem we should implement  spinlock (unfortunately I don't think there is one in C++11, but it does provide the [primitives](https://en.cppreference.com/w/cpp/atomic/atomic_flag) necessary)

Reply:
Sure，let me go ahead and have the access under lock. We can revisit this if it the performance turns out to be an issue.